### PR TITLE
add: expense tracker tutorial

### PR DESCRIPTION
**Pull request by GitBolt**
Tutorial for Expense Tracker using PDAs for the upcoming Solana Workshop https://github.com/Solana-Workshops/.contributing/issues/2

### CODE REVIEW

File: client/src/tutorials/ExpenseTracker/files/index.test.ts.raw
Comment (acheroncrypto) [inline client/src/tutorials/ExpenseTracker/files/index.test.ts.raw:1]: Have you tested this in playground?

File: client/src/tutorials/ExpenseTracker/files/lib.rs
Comment (acheroncrypto) [inline client/src/tutorials/ExpenseTracker/files/lib.rs:1]: Could you format this file with rustfmt?

File: client/src/tutorials/ExpenseTracker/files/lib.rs [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    #[account(
+        init,
+        payer = authority,
+        space = 8 
+            + 8 // id
+            + 32 // owner
+            + (4 + 12) // merchant name
+            + 8 // amount
+            + 1, // bump
+        seeds = [b"expense".as_ref(),authority.key().as_ref(),id.as_ref()],
```
Comment (acheroncrypto) [inline client/src/tutorials/ExpenseTracker/files/lib.rs:68]: Byte string `b"expense"` is already a reference, `.as_ref()` is redundant, and some spacing would help.
```suggestion
        seeds = [b"expense", authority.key().as_ref(), id.as_ref()], 
```

File: client/src/tutorials/ExpenseTracker/files/lib.rs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+            + (4 + 12) // merchant name
+            + 8 // amount
+            + 1, // bump
+        seeds = [b"expense".as_ref(),authority.key().as_ref(),id.as_ref()], 
+        bump
+    )]
+    pub expense_account: Account<'info, ExpenseAccount>,
+
+    // Misc Accounts
+    #[account(address = system_program::ID)]
```
Comment (acheroncrypto) [inline client/src/tutorials/ExpenseTracker/files/lib.rs:74]: Redundant check, `pub system_program: Program<'info,System>,` already does this check.
```suggestion
```

File: client/src/tutorials/ExpenseTracker/files/lib.rs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    pub system_program: Program<'info,System>,
+    #[account(address = solana_program::sysvar::rent::ID)]
+    pub rent: Sysvar<'info, Rent>,
```
Comment (acheroncrypto) [inline client/src/tutorials/ExpenseTracker/files/lib.rs:77]: Rent is not necessary to pass as an account, you can get it with `Rent::get()?` inside the program which decreases transaction size and is potentially more secure.

File: client/src/tutorials/ExpenseTracker/files/lib.rs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        let integer_id = id.parse::<u64>().unwrap();
+        let string_id = integer_id.to_string();
+
+        require_eq!(string_id, id);
+
+        expense_account.id = integer_id;
+        expense_account.mname = mname;
+        expense_account.amount = amount;
+        expense_account.owner = *ctx.accounts.authority.key;
+        expense_account.bump = *ctx.bumps.get("expense_account").unwrap();
```
Comment (acheroncrypto) [inline client/src/tutorials/ExpenseTracker/files/lib.rs:26]: Saving bumps is not necessary, using `#[account(bump, ...)]` will get the same result.

**Review by acheroncrypto — CHANGES_REQUESTED** [review]
Thank you. Could you please format the code with `rustfmt` and generally follow Rust's guidelines? It's much better to have a standardized format for tutorials.

I've also noted some of the redundant checks and please make sure to test this and confirm it works as expected in playground.

**Comment by GitBolt** [comment]
Appreciate the detailed review! Fixing these and getting back you

**Comment by GitBolt** [comment]
hey @acheroncrypto, i've made the changes. please have a look! i've fixed tests, formatted code, removed the unnecessary stuff and checked everything by running in playground

File: client/src/tutorials/ExpenseTracker/files/lib.rs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+declare_id!("");
+
+#[program]
+pub mod etracker {
+
+    use super::*;
+
+    pub fn initialize_expense(
+        ctx: Context<InitializeExpense>,
+        id: String,
```
Comment (acheroncrypto) [inline client/src/tutorials/ExpenseTracker/files/lib.rs:13]: Why pass a `String` here if the program panics for values that are not parseable to `u64`?

```suggestion
        id: u64,
```

File: client/src/tutorials/ExpenseTracker/files/lib.rs [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        merchant_name: String,
+        amount: u64,
+    ) -> Result<()> {
+        let expense_account = &mut ctx.accounts.expense_account;
+        let integer_id = id.parse::<u64>().unwrap();
+        let string_id = integer_id.to_string();
+
+        require_eq!(string_id, id);
```
Comment (acheroncrypto) [inline client/src/tutorials/ExpenseTracker/files/lib.rs:21]: We are parsing the `String` to `u64` and then we are converting the parsed `u64` value to `String`. Lastly, we compare the two `String` values to be equal.

```suggestion
```

File: client/src/tutorials/ExpenseTracker/files/index.test.ts.raw [hunk 7]
```diff
@@ -0,0 +1,45 @@
+describe("Expense Tracker", async () => {
+  let merchant_name = "test";
```
Comment (acheroncrypto) [inline client/src/tutorials/ExpenseTracker/files/index.test.ts.raw:2]: Can we use camelCase naming in TypeScript files for consistency?

File: client/src/tutorials/ExpenseTracker/pages/5.md [hunk 8]
```diff
@@ -0,0 +1,48 @@
+# Fetching Expense Entries
+Now you may be wondering, how do I fetch all expense entries owned by a particular user?
+
+Anchor provides us .all() and .fetch() methods. One returns all expense accounts, other returns an individual fetch account through its public key. But what about adding some filtering to fetch accounts by some specific attribute, like the owner?
+
+Turns out, it's super easy!
+Have a look into this snippet:
+
+```ts
+const expenses = await program.account.expenseAccount.all([
```
Comment (acheroncrypto) [inline client/src/tutorials/ExpenseTracker/pages/5.md:10]:
```suggestion
const expenses = await pg.program.account.expenseAccount.all([
```

File: client/src/tutorials/ExpenseTracker/pages/5.md [hunk 9]
```diff
@@ -0,0 +1,48 @@
+# Fetching Expense Entries
+Now you may be wondering, how do I fetch all expense entries owned by a particular user?
+
+Anchor provides us .all() and .fetch() methods. One returns all expense accounts, other returns an individual fetch account through its public key. But what about adding some filtering to fetch accounts by some specific attribute, like the owner?
+
+Turns out, it's super easy!
+Have a look into this snippet:
+
+```ts
```
Comment (acheroncrypto) [inline client/src/tutorials/ExpenseTracker/pages/5.md:9]: Can we format this snippet?

File: client/src/tutorials/ExpenseTracker/files/lib.rs [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+        seeds = [b"expense", authority.key().as_ref(), id.as_ref()], 
+        bump
+    )]
+    pub expense_account: Account<'info, ExpenseAccount>,
+
+    pub system_program: Program<'info, System>,
+}
+
+#[derive(Accounts)]
+#[instruction(id : String)]
```
Comment (acheroncrypto) [inline client/src/tutorials/ExpenseTracker/files/lib.rs:84]:
```suggestion
#[instruction(id: String)]
```

**Review by acheroncrypto — COMMENTED** [review]
Thank you for fixing the earlier review comments, it looks much better now.

File: client/src/tutorials/ExpenseTracker/files/lib.rs (hunk 5 above)
Comment (GitBolt) [reply client/src/tutorials/ExpenseTracker/files/lib.rs:13]: I use string cause the id is seed as well and it's easier to process strings

**Comment by GitBolt** [comment]
I've fixed other issues, do let me know your thoughts on the string/u64 one in the instruction

File: client/src/tutorials/ExpenseTracker/files/lib.rs (hunk 5 above)
Comment (acheroncrypto) [reply client/src/tutorials/ExpenseTracker/files/lib.rs:13]: I don't think it's justified. Clients should adjust based on the on-chain program, not the other way around.

Comment (GitBolt) [reply client/src/tutorials/ExpenseTracker/files/lib.rs]: Updated, can you check now?

File: client/src/tutorials/ExpenseTracker/files/lib.rs (hunk 6 above)
Comment (GitBolt) [reply client/src/tutorials/ExpenseTracker/files/lib.rs]: Removed

File: client/src/tutorials/ExpenseTracker/pages/5.md [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+const expenses = await pg.program.account.expenseAccount.all([
+  {
+    memcmp: {
+      offset: 8 + 8,
+      bytes: wallet.publicKey.toBase58(),
```
Comment (acheroncrypto) [inline client/src/tutorials/ExpenseTracker/pages/5.md:14]:
```suggestion
      bytes: pg.wallet.publicKey.toBase58(),
```

**Comment by acheroncrypto** [comment]
Looking good overall but CI is failing due to formatting.

**Comment by GitBolt** [comment]
Oh weird, I did format the code through Solpg. Let me check again

**Comment by acheroncrypto** [comment]
> Oh weird, I did format the code through Solpg. Let me check again

You can run `yarn format` to fix it.

**Comment by GitBolt** [comment]
> > Oh weird, I did format the code through Solpg. Let me check again
> 
> You can run `yarn format` to fix it.

Oh awesome, thanks! Fixed it. Thought tutorial code examples were failing

**Comment by GitBolt** [comment]
Is there anything left on this or its good to merge now? @acheroncrypto

**Review by acheroncrypto — APPROVED** [review]
Thank you for your contribution! I'm merging it now but it will take some time to show inside the app.

**Pull request merged**